do not force to set setTitles() in a Builder

> from https://github.com/dvdciri/MultiChoiceRecyclerView/issues/28#issuecomment-274800577

if needed I think the constructor would be a better place

The `MultiChoiceToolbar#Builder()` takes `setTitles()`.
If I **do not** add this to the builder, it crashes with:

```java
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String java.lang.String.trim()' on a null object reference
at com.davidecirillo.multichoicerecyclerview.MultiChoiceToolbar.<init>(MultiChoiceToolbar.java:26)
at com.davidecirillo.multichoicerecyclerview.MultiChoiceToolbar.<init>(MultiChoiceToolbar.java:9)
at com.davidecirillo.multichoicerecyclerview.MultiChoiceToolbar$Builder.build(MultiChoiceToolbar.java:122)
```

In my understanding, a builder is the thing where I *might* add options (but I *can* leave them unset).
Am I wrong?

---

This is not a bug - more like a architectural discussion ;-)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [Unreleased]
 ### Added
+- Possibility to use QuantityStrings for the toolbar while selecting items
+
 ### Changed
 ### Fixed
 ### Deleted

--- a/library/multichoicerecyclerview/build.gradle
+++ b/library/multichoicerecyclerview/build.gradle
@@ -8,7 +8,7 @@ checkstyle {
 }
 
 ext {
-    PUBLISH_VERSION = '2.0.0'
+    PUBLISH_VERSION = '2.1.0'
     SUPPORT_LIBRARY_VERSION = '24.2.1'
     BUILD_TOOLS = "23.0.3"
     TARGET_SDK = 23

--- a/library/multichoicerecyclerview/src/main/java/com/davidecirillo/multichoicerecyclerview/Constants.java
+++ b/library/multichoicerecyclerview/src/main/java/com/davidecirillo/multichoicerecyclerview/Constants.java
@@ -1,0 +1,10 @@
+package com.davidecirillo.multichoicerecyclerview;
+
+public class Constants {
+
+    /**
+     * Used for res-variables as a default value.
+     * This way, we can detect if a value was set or not.
+     */
+    public static final int INVALID_RES = -1;
+}

--- a/library/multichoicerecyclerview/src/main/java/com/davidecirillo/multichoicerecyclerview/MultiChoiceToolbar.java
+++ b/library/multichoicerecyclerview/src/main/java/com/davidecirillo/multichoicerecyclerview/MultiChoiceToolbar.java
@@ -1,6 +1,7 @@
 package com.davidecirillo.multichoicerecyclerview;
 
 import android.content.res.TypedArray;
+import android.support.annotation.PluralsRes;
 import android.support.v7.app.AppCompatActivity;
 import android.support.v7.widget.Toolbar;
 import android.util.TypedValue;
@@ -11,6 +12,7 @@ public class MultiChoiceToolbar {
     private AppCompatActivity mAppCompatActivity;
     private Toolbar mToolbar;
     private String mDefaultToolbarTitle;
+    private @PluralsRes int mSelectedQuantityTitle;
     private String mSelectedToolbarTitle;
     private int mDefaultPrimaryColor = 0;
     private int mDefaultPrimaryColorDark = 0;
@@ -24,6 +26,7 @@ public class MultiChoiceToolbar {
         this.mAppCompatActivity = builder.mAppCompatActivity;
         this.mToolbar = builder.mToolbar;
         this.mDefaultToolbarTitle = builder.mDefaultToolbarTitle.trim();
+        this.mSelectedQuantityTitle = builder.mSelectedQuantityTitle;
         this.mSelectedToolbarTitle = builder.mSelectedToolbarTitle.trim();
         this.mDefaultPrimaryColor = builder.mDefaultPrimaryColor;
         this.mDefaultPrimaryColorDark = builder.mDefaultPrimaryColorDark;
@@ -43,8 +46,9 @@ public class MultiChoiceToolbar {
     public static class Builder {
         private AppCompatActivity mAppCompatActivity;
         private Toolbar mToolbar;
-        private String mDefaultToolbarTitle;
-        private String mSelectedToolbarTitle;
+        private String mDefaultToolbarTitle = "";
+        private @PluralsRes int mSelectedQuantityTitle = Constants.INVALID_RES;
+        private String mSelectedToolbarTitle = "";
 
         // Colours
         private int mDefaultPrimaryColor = 0;
@@ -101,6 +105,19 @@ public class MultiChoiceToolbar {
         }
 
         /**
+         * Set the titles for the differente state of the toolbar default/multiChoice
+         *
+         * @param defaultTitle          Title when the toolbar shown is the default one
+         * @param selectedQuantityTitle Title shown when some item are selected. Will use the plural-feature to let you define the correct format.
+         * @return Builder so you can chain together setters and build
+         */
+        public Builder setTitles(String defaultTitle, @PluralsRes int selectedQuantityTitle) {
+            this.mDefaultToolbarTitle = defaultTitle;
+            this.mSelectedQuantityTitle = selectedQuantityTitle;
+            return this;
+        }
+
+        /**
          * Set the default icon that will be shown when is not in multi choice mode.
          * If not set there will be no icon.
          *
@@ -143,6 +160,11 @@ public class MultiChoiceToolbar {
 
     String getDefaultToolbarTitle() {
         return mDefaultToolbarTitle;
+    }
+
+    @PluralsRes
+    int getSelectedToolbarQuantityTitle() {
+        return mSelectedQuantityTitle;
     }
 
     String getSelectedToolbarTitle() {

--- a/library/multichoicerecyclerview/src/main/java/com/davidecirillo/multichoicerecyclerview/MultiChoiceToolbarHelper.java
+++ b/library/multichoicerecyclerview/src/main/java/com/davidecirillo/multichoicerecyclerview/MultiChoiceToolbarHelper.java
@@ -2,9 +2,11 @@ package com.davidecirillo.multichoicerecyclerview;
 
 import android.graphics.drawable.ColorDrawable;
 import android.os.Build;
+import android.support.annotation.PluralsRes;
 import android.support.v4.content.ContextCompat;
 import android.support.v7.app.ActionBar;
 import android.support.v7.app.AppCompatActivity;
+import android.text.TextUtils;
 import android.view.View;
 import android.view.Window;
 import android.view.WindowManager;
@@ -94,12 +96,20 @@ class MultiChoiceToolbarHelper {
 
     private void updateToolbarTitle(int itemNumber) {
         if (itemNumber > 0) {
+            @PluralsRes int selectedToolbarQuantityTitle = mMultiChoiceToolbar.getSelectedToolbarQuantityTitle();
             String selectedToolbarTitle = mMultiChoiceToolbar.getSelectedToolbarTitle();
-            if (selectedToolbarTitle != null) {
+
+            // if set, prefer the quantity-res over the given string or just use the number of selected items
+            if (selectedToolbarQuantityTitle != 0 && selectedToolbarQuantityTitle != Constants.INVALID_RES) {
+                mSupportActionBar.setTitle(mAppCompatActivity.getResources().getQuantityString(selectedToolbarQuantityTitle, itemNumber, itemNumber));
+
+            } else if (selectedToolbarTitle != null) {
                 mSupportActionBar.setTitle(String.format(Locale.UK, SELECTED_TOOLBAR_TITLE_FORMAT, itemNumber, selectedToolbarTitle));
+
             } else {
-                mSupportActionBar.setTitle(itemNumber);
+                mSupportActionBar.setTitle(String.valueOf(itemNumber));
             }
+
         } else {
             String defaultToolbarTitle = mMultiChoiceToolbar.getDefaultToolbarTitle();
             if (defaultToolbarTitle != null) {

--- a/sample/src/androidTest/java/com/davidecirillo/multichoicesample/SampleToolbarActivityPluralsTest.java
+++ b/sample/src/androidTest/java/com/davidecirillo/multichoicesample/SampleToolbarActivityPluralsTest.java
@@ -1,0 +1,127 @@
+package com.davidecirillo.multichoicesample;
+
+import android.graphics.drawable.ColorDrawable;
+import android.support.test.espresso.contrib.RecyclerViewActions;
+import android.support.test.rule.ActivityTestRule;
+import android.support.test.runner.AndroidJUnit4;
+import android.support.v4.content.ContextCompat;
+import android.support.v7.widget.Toolbar;
+import android.view.View;
+import android.widget.RelativeLayout;
+import android.widget.TextView;
+
+import com.davidecirillo.multichoicesample.api.BaseMultiChoiceActivityTest;
+import com.davidecirillo.multichoicesample.sampleToolbar.SampleToolbarActivity;
+
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import static android.support.test.InstrumentationRegistry.getInstrumentation;
+import static android.support.test.espresso.Espresso.onView;
+import static android.support.test.espresso.Espresso.openActionBarOverflowOrOptionsMenu;
+import static android.support.test.espresso.action.ViewActions.click;
+import static android.support.test.espresso.action.ViewActions.longClick;
+import static android.support.test.espresso.assertion.ViewAssertions.matches;
+import static android.support.test.espresso.matcher.ViewMatchers.isAssignableFrom;
+import static android.support.test.espresso.matcher.ViewMatchers.withId;
+import static android.support.test.espresso.matcher.ViewMatchers.withParent;
+import static android.support.test.espresso.matcher.ViewMatchers.withText;
+import static org.hamcrest.Matchers.allOf;
+import static org.hamcrest.core.StringContains.containsString;
+
+@RunWith(AndroidJUnit4.class)
+public class SampleToolbarActivityPluralsTest extends BaseMultiChoiceActivityTest {
+
+    public SampleToolbarActivity mActivity;
+
+    // region BaseMultiChoiceActivityTest
+
+    @Override
+    protected boolean isSelected(View view) {
+        RelativeLayout relativeLayout = (RelativeLayout) view.findViewById(R.id.get_started_relative_layout);
+
+        return !(relativeLayout == null || relativeLayout.getBackground() == null) && ((ColorDrawable) relativeLayout.getBackground()).getColor() == ContextCompat.getColor(mActivity, R.color.colorPrimaryDark);
+    }
+
+    // endregion
+
+    // region tests
+
+    @Rule
+    public ActivityTestRule<SampleToolbarActivity> mActivityRule = new ActivityTestRule<>(SampleToolbarActivity.class);
+
+    @Before
+    public void setUp() {
+        mActivity = mActivityRule.getActivity();
+
+        mActivity.runOnUiThread(new Runnable() {
+            public void run() {
+                mActivity.setQuantityMode(SampleToolbarActivity.QuantityMode.PLURALS);
+            }
+        });
+
+        wakeScreen(mActivity);
+    }
+
+    /**
+     * #5 testMultiChoiceToolbar
+     * <p>
+     * Make sure that the multi choice toolbar is displayed and updated with the correct number of selected items
+     */
+    @Test
+    public void testMultiChoiceToolbarPlurals() {
+        onView(withId(R.id.multiChoiceRecyclerView))
+                .perform(RecyclerViewActions.actionOnItemAtPosition(0, longClick()));
+
+        onView(allOf(isAssignableFrom(TextView.class), withParent(isAssignableFrom(Toolbar.class))))
+                .check(matches(withText(containsString("One"))));
+
+        onView(withId(R.id.multiChoiceRecyclerView))
+                .perform(RecyclerViewActions.actionOnItemAtPosition(1, click()));
+
+        onView(allOf(isAssignableFrom(TextView.class), withParent(isAssignableFrom(Toolbar.class))))
+                .check(matches(withText(containsString("2"))));
+
+        onView(withId(R.id.multiChoiceRecyclerView))
+                .perform(RecyclerViewActions.actionOnItemAtPosition(1, click()));
+
+        onView(withId(R.id.multiChoiceRecyclerView))
+                .perform(RecyclerViewActions.actionOnItemAtPosition(0, click()));
+
+        onView(allOf(isAssignableFrom(TextView.class), withParent(isAssignableFrom(Toolbar.class))))
+                .check(matches(withText(R.string.toolbar_controls)));
+    }
+
+    /**
+     * #9 testSingleClickModeOptionMenu
+     * <p>
+     * Find and Click on the option menu item "Single Click Mode" and check that the functionality is activated on the
+     * MultiChoiceRecyclerView
+     */
+    @Test
+    public void testSingleClickModeOptionMenu() {
+        openActionBarOverflowOrOptionsMenu(getInstrumentation().getTargetContext());
+
+        onView(withText("Single Click Mode"))
+                .perform(click());
+
+        onView(withId(R.id.multiChoiceRecyclerView))
+                .perform(RecyclerViewActions.actionOnItemAtPosition(0, click()))
+                .check(matches(isSelected()));
+
+        //Check the toolbar as well
+        onView(allOf(isAssignableFrom(TextView.class), withParent(isAssignableFrom(Toolbar.class))))
+                .check(matches(withText(containsString("One"))));
+
+        onView(withId(R.id.multiChoiceRecyclerView))
+                .perform(RecyclerViewActions.actionOnItemAtPosition(1, click()))
+                .check(matches(isSelected()));
+
+        onView(allOf(isAssignableFrom(TextView.class), withParent(isAssignableFrom(Toolbar.class))))
+                .check(matches(withText(containsString("2"))));
+    }
+
+    // endregion
+}

--- a/sample/src/main/java/com/davidecirillo/multichoicesample/sampleToolbar/SampleToolbarActivity.java
+++ b/sample/src/main/java/com/davidecirillo/multichoicesample/sampleToolbar/SampleToolbarActivity.java
@@ -184,6 +184,8 @@ public class SampleToolbarActivity extends BaseActivity {
 
                 case R.id.plural_mode:
                     setQuantityMode(QuantityMode.PLURALS);
+
+                    Toast.makeText(getApplicationContext(), "set toolbar to use QuantityMode.PLURALS", Toast.LENGTH_SHORT).show();
                     return true;
             }
         }

--- a/sample/src/main/java/com/davidecirillo/multichoicesample/sampleToolbar/SampleToolbarActivity.java
+++ b/sample/src/main/java/com/davidecirillo/multichoicesample/sampleToolbar/SampleToolbarActivity.java
@@ -2,6 +2,7 @@ package com.davidecirillo.multichoicesample.sampleToolbar;
 
 import android.content.Intent;
 import android.os.Bundle;
+import android.support.annotation.IntDef;
 import android.support.annotation.Nullable;
 import android.support.v7.widget.GridLayoutManager;
 import android.support.v7.widget.LinearLayoutManager;
@@ -16,6 +17,8 @@ import com.davidecirillo.multichoicesample.BaseActivity;
 import com.davidecirillo.multichoicesample.R;
 import com.davidecirillo.multichoicesample.ResultActivity;
 
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
 import java.util.ArrayList;
 import java.util.List;
 
@@ -29,12 +32,27 @@ import rx.schedulers.Schedulers;
 
 public class SampleToolbarActivity extends BaseActivity {
 
+    private static final int DEFAULT_QUANTITY_MODE = QuantityMode.STRING;
+
+    @IntDef({
+            QuantityMode.NONE,
+            QuantityMode.STRING,
+            QuantityMode.PLURALS,
+    })
+    @Retention(RetentionPolicy.SOURCE)
+    public @interface QuantityMode {
+        int NONE = 0;
+        int STRING = 1;
+        int PLURALS = 2;
+    }
+
     public static final String SELECTED_ITEMS = "selectedItems";
     @BindView(R.id.multiChoiceRecyclerView)
     RecyclerView mMultiChoiceRecyclerView;
 
     private ArrayList<String> stringList;
     private MySampleToolbarAdapter mMySampleToolbarAdapter;
+    private @SampleToolbarActivity.QuantityMode int quantityMode = DEFAULT_QUANTITY_MODE;
 
     @Override
     protected void onCreate(@Nullable Bundle savedInstanceState) {
@@ -88,17 +106,33 @@ public class SampleToolbarActivity extends BaseActivity {
     private void setUpAdapter(List<String> strings) {
         stringList = new ArrayList<>(strings);
 
-        MultiChoiceToolbar multiChoiceToolbar =
-                new MultiChoiceToolbar.Builder(SampleToolbarActivity.this, toolbar)
-                        .setTitles(toolbarTitle(), "item selected")
-                        .setMultiChoiceColours(R.color.colorPrimaryMulti, R.color.colorPrimaryDarkMulti)
-                        .setDefaultIcon(R.drawable.ic_arrow_back_white_24dp, new View.OnClickListener() {
-                            @Override
-                            public void onClick(View view) {
-                                onBackPressed();
-                            }
-                        })
-                        .build();
+        MultiChoiceToolbar.Builder builder = new MultiChoiceToolbar.Builder(SampleToolbarActivity.this, toolbar)
+                .setMultiChoiceColours(R.color.colorPrimaryMulti, R.color.colorPrimaryDarkMulti)
+                .setDefaultIcon(R.drawable.ic_arrow_back_white_24dp, new View.OnClickListener() {
+                    @Override
+                    public void onClick(View view) {
+                        onBackPressed();
+                    }
+                });
+
+        switch (quantityMode) {
+            case QuantityMode.NONE:
+                /*
+                 * You may use one of the setTitles()-methods
+                 * or none (it will then show the count of selected items instead).
+                 */
+                break;
+
+            case QuantityMode.STRING:
+                builder.setTitles(toolbarTitle(), "item selected");
+                break;
+
+            case QuantityMode.PLURALS:
+                builder.setTitles(toolbarTitle(), R.plurals.numberOfSelectedItems);
+                break;
+        }
+
+        MultiChoiceToolbar multiChoiceToolbar = builder.build();
 
         mMySampleToolbarAdapter = new MySampleToolbarAdapter(stringList, getApplicationContext());
         mMySampleToolbarAdapter.setMultiChoiceToolbar(multiChoiceToolbar);
@@ -126,26 +160,34 @@ public class SampleToolbarActivity extends BaseActivity {
 
     @Override
     public boolean onOptionsItemSelected(MenuItem item) {
-
         if (mMySampleToolbarAdapter != null) {
             switch (item.getItemId()) {
                 case R.id.select_all:
                     mMySampleToolbarAdapter.selectAll();
                     return true;
+
                 case R.id.deselect_all:
                     mMySampleToolbarAdapter.deselectAll();
                     return true;
+
                 case R.id.select_3:
                     mMySampleToolbarAdapter.select(2);
                     return true;
-                case R.id.single_click_mode:
 
+                case R.id.single_click_mode:
                     mMySampleToolbarAdapter.setSingleClickMode(!mMySampleToolbarAdapter.isInSingleClickMode());
-                    Toast.makeText(getApplicationContext(), "Always Single Click Mode [" + mMySampleToolbarAdapter.isInSingleClickMode() + "]", Toast
-                            .LENGTH_SHORT).show();
+                    Toast.makeText(getApplicationContext(),
+                            "Always Single Click Mode [" + mMySampleToolbarAdapter.isInSingleClickMode() + "]",
+                            Toast.LENGTH_SHORT).show();
+
+                    return true;
+
+                case R.id.plural_mode:
+                    setQuantityMode(QuantityMode.PLURALS);
                     return true;
             }
         }
+
         return false;
     }
 
@@ -157,5 +199,17 @@ public class SampleToolbarActivity extends BaseActivity {
     @Override
     protected String toolbarTitle() {
         return getString(R.string.toolbar_controls);
+    }
+
+    @QuantityMode
+    public int getQuantityMode() {
+        return quantityMode;
+    }
+
+    public void setQuantityMode(@QuantityMode int quantityMode) {
+        mMySampleToolbarAdapter.deselectAll();
+
+        this.quantityMode = quantityMode;
+        setUpMultiChoiceRecyclerView();
     }
 }

--- a/sample/src/main/res/menu/sample_toolbar_menu.xml
+++ b/sample/src/main/res/menu/sample_toolbar_menu.xml
@@ -25,4 +25,10 @@
         android:title="@string/single_click_mode"
         app:showAsAction="never"
         />
+
+    <item
+        android:id="@+id/plural_mode"
+        android:title="@string/plural_mode"
+        app:showAsAction="never"
+        />
 </menu>

--- a/sample/src/main/res/values/strings.xml
+++ b/sample/src/main/res/values/strings.xml
@@ -17,4 +17,10 @@
     <string name="deselect_all">Deselect All</string>
     <string name="select_third_item">Select Third Item</string>
     <string name="single_click_mode">Single Click Mode</string>
+    <string name="plural_mode">Plural Mode</string>
+
+    <plurals name="numberOfSelectedItems">
+        <item quantity="one">One item selected</item>
+        <item quantity="other">%d items selected</item>
+    </plurals>
 </resources>

